namerd/iface/control-http/src/main/scala/io/buoyant/namerd/iface/HttpControlService.scala

```diff
@@ -170,26 +171,75 @@ class HttpControlService(storage: DtabStore, delegate: Ns => NameInterpreter, na
       resp
   }
 
-  private[this] def handleList(): Future[Response] =
-    storage.list().map { namespaces =>
-      val rsp = Response()
-      rsp.contentType = MediaType.Json
-      rsp.content = Json.write(namespaces)
-      rsp
+  private[this] def streamingResp[T](
+    values: Event[Try[T]],
+    contentType: Option[String] = None
+  )(render: T => Buf): Future[Response] = {
+    val resp = Response()
+    for (ct <- contentType) resp.contentType = ct
+    resp.setChunked(true)
+    val writer = resp.writer
+    // closable is a handle to the values observation so that we can close the observation when the
+    // streaming connection is terminated
+    @volatile var closable: Closable = Closable.nop
+    // calls to writer.write must be flatMapped together to ensure proper ordering and backpressure
+    // writeFuture is an accumulator of those flatMapped Futures
+    @volatile var writeFuture: Future[Unit] = Future.Unit
+    closable = values.respond {
+      case Return(t) =>
+        writeFuture = writeFuture.before {
+          val buf = render(t)
+          if (buf == Buf.Empty)
+            Future.Unit
+          else
+            writer.write(buf).onFailure { _ => closable.close() }
+        }
+      case Throw(e) =>
+        val _ = writer.write(Buf.Utf8(e.getMessage).concat(newline)).before(writer.close())
```

Comment:
In general, that seems reasonable.  In our case we don't stream back any json and each render method would end up having the same handling for Throw (ie just print the exception message).  I think we can maybe defer that until we are returning content types other than plain text.